spec.html

```diff
@@ -28653,9 +28661,9 @@ <h2>Syntax</h2>
           `export` ExportFromClause FromClause `;`
           `export` NamedExports `;`
           `export` VariableStatement[~Yield, +Await]
-          `export` Declaration[~Yield, +Await]
+          DecoratorList[~Yield, +Await]? `export` Declaration[~Yield, +Await]
```

Comment:
I think it would be slightly easier to follow the grammar if it was defined like this:
```js
`export` LexicalDeclaration[~Yield, +Await]
`export` HoistableDeclaration[~Yield, +Await]
`export` ClassDeclaration[~Yield, +Await, +Decorators]
DecoratorList[~Yield, +Await]? `export` ClassDeclaration[~Yield, +Await, ~Decorators]

`export` `default` HoistableDeclaration[~Yield, +Await, +Default]
`export` `default` ClassDeclaration[~Yield, +Await, +Decorators, +Default]
DecoratorList[~Yield, +Await]? `export` `default` ClassDeclaration[~Yield, +Await, +Decorators, ~Default]
`export` `default` [lookahead &notin; { `function`, `async` [no LineTerminator here] `function`, `class`, `@` }] AssignmentExpression[+In, ~Yield, +Await] `;`
```

because the grammar rules are in a single place and you don't have to read both the grammar and the syntactic early errors. However, the end result is effectively the same so avoiding the new production parameter as you proposed is also ok :+1:

Reply:
I considered that, and that was what I'd implemented originally in https://github.com/tc39/proposal-decorators/pull/95/files. The early error seemed a little easier to reason over given that a `Decorators` parameter doesn't go very deep.

Reply:
There's also a [bug in `esmeta`](https://github.com/es-meta/esmeta/issues/133) regarding multiple constraints on an RHS as would be needed here. That bug has since been fixed, but hasn't been released yet. I think there's still some work to go on the decorators PR before it will pass `esmeta` anyways, so it doesn't really matter too much either way.

Reply:
I think I would prefer to have a new `UndecoratedClassDeclaration` nonterminal which derives what `ClassDeclaration` currently does in ecma262, and then have `ClassDeclaration` derive `DecoratorList? UndecoratedClassDeclaration`. Then you can use `UndecoratedClassDeclaration` here, no parameters or early errors necessary.

Still, that's an editorial change which can be made later. This change is correct as-is. Only reason to do it now is if it would cause TS or Babel to use a different representation in their ASTs; I know sometimes people like to have the AST shapes stick close to the spec (though this seems like a case where you would not want that).

Reply:
I don't know about Babel, but TypeScript won't introduce a new AST node to differentiate. We just store decorators as if they were modifiers now (a change we made in TS 4.9 to support decorators-after-`export` syntax), so `export`, `default`, `declare`, `abstract`, and decorators all end up in the same bucket. An `UndecoratedClassDeclaration` production feels more like a workaround for the grammar, much like cover grammars. I don't expect we would really tie any unique semantics to it, and it might even make handling `Function.prototype.toString` a bit more complicated.

Reply:
Eh, I disagree - using productions which precisely express what you actually want to accept is the normal use of a CFG, and grammar parameters and early errors are the workaround for when that's not practical.

Fair point about `Function.prototype.toString`, but I'd have to actually write it out to tell if it will actually be a problem.